pkg/loop/plugin_service.go

```diff
@@ -0,0 +1,206 @@
+package loop
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os/exec"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/go-plugin"
+	"golang.org/x/exp/maps"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/smartcontractkit/chainlink-relay/pkg/logger"
+	"github.com/smartcontractkit/chainlink-relay/pkg/types"
+	"github.com/smartcontractkit/chainlink-relay/pkg/utils"
+)
+
+type grpcPlugin interface {
+	plugin.Plugin
+	plugin.GRPCPlugin
+	ClientConfig() *plugin.ClientConfig
+}
+
+// RelayerService is a [types.Service] wrapper that maintains an internal [types.Service] created from a [GRPCPlugin]
+// client instance by launching and re-launching as necessary.
+type pluginService[P grpcPlugin, S types.Service] struct {
+	utils.StartStopOnce
+
+	pluginName string
+
+	lggr logger.Logger
+	cmd  func() *exec.Cmd
+
+	wg     sync.WaitGroup
+	stopCh chan struct{}
+
+	grpcPlug P
+
+	client         *plugin.Client
+	clientProtocol plugin.ClientProtocol
+
+	newService func(context.Context, any) (S, error)
+
+	serviceCh chan struct{} // closed when service is available
+	service   S
+
+	testInterrupt chan func(*pluginService[P, S]) // tests only (via TestHook) to enable access to internals without racing
+}
+
+func newPluginService[P grpcPlugin, S types.Service](pluginName string, p P, newService func(context.Context, any) (S, error), lggr logger.Logger, cmd func() *exec.Cmd, stopCh chan struct{}) *pluginService[P, S] {
+	return &pluginService[P, S]{
+		pluginName: pluginName,
+		lggr:       lggr,
+		cmd:        cmd,
+		stopCh:     stopCh,
+		grpcPlug:   p,
+		newService: newService,
+		serviceCh:  make(chan struct{}),
+	}
+}
+
+func (r *pluginService[P, S]) keepAlive() {
+	defer r.wg.Done()
+
+	r.lggr.Debugw("Staring keepAlive", "tick", keepAliveTickDuration)
+
+	t := time.NewTicker(keepAliveTickDuration)
+	defer t.Stop()
+	for {
+		select {
+		case <-r.stopCh:
+			return
+		case <-t.C:
+			c := r.client
+			cp := r.clientProtocol
+			if c != nil && !c.Exited() && cp != nil {
+				// launched
+				err := cp.Ping()
+				if err == nil {
+					continue // healthy
+				}
+				r.lggr.Errorw("Relaunching unhealthy plugin", "err", err)
+			}
+			if err := r.tryLaunch(cp); err != nil {
+				r.lggr.Errorw("Failed to launch plugin", "err", err)
+			}
+		case fn := <-r.testInterrupt:
+			fn(r)
+		}
+	}
+}
+
+func (r *pluginService[P, S]) tryLaunch(old plugin.ClientProtocol) (err error) {
+	if old != nil && r.clientProtocol != old {
+		// already replaced by another routine
+		return nil
+	}
+	r.client, r.clientProtocol, err = r.launch()
+	return
+}
+
+func (r *pluginService[P, S]) launch() (*plugin.Client, plugin.ClientProtocol, error) {
+	ctx, cancelFn := utils.ContextFromChan(r.stopCh)
+	defer cancelFn()
+
+	r.lggr.Debug("Launching")
+
+	cc := r.grpcPlug.ClientConfig()
+	cc.Cmd = r.cmd()
+	client := plugin.NewClient(cc)
+	cp, err := client.Client()
```

Comment:
why is a ClientProtoocal returned from client.Client()?

is this our naming or hashicorp?

aside: what is the difference between a Client and a client protocol?

Reply:
It's their naming. It feels like there's an extra type, but I don't fully grok it. Will revisit.